### DISCUSSION TITLE
How do I explain classic time series models such as ARIMA, FBProhphet etc., is there a way to do that in SHAP?

### DISCUSSION THREAD

**Discussion by shbhmpndy** [shap/shap#1464] [General]
I was looking into deep learning based time series models for explanation, but couldn't find anything for explaining models like ARIMA, FBProphet etc.
I have two areas of concern to tackle
1. Can I treat the time series problem as a regression problem and explain the exogenous features? Will that make sense? Also if there are no exogenous features, will saying that "so and so time frame was responsible for the prediction" make sense?
2. Following from the above question, does including t-1, t-2.. as features and explaining them do any good? So for example in case of weekly data you can say something like "Your current week forecast is X because the values in N week were so and so"?

**Comment by Kamalnor** [comment]
Following

**Comment by slundberg** [comment]
The short answer is that I think you should explain the impact of past time points, since that is often the biggest driving factor of the prediction. The tricky part about time series data is the strong time dependence makes realistically masking the features tricky. While it would be great to have, there are no time-series maskers built into SHAP right now. But you could use a custom masker and just impute forward or something. See here for an example of using a custom masking function in the new API: https://github.com/slundberg/shap/blob/master/notebooks/maskers/custom.ipynb

You could even go more all out and try to build a masker that is based on higher level features of the time series, like a DCT, or FWT transform. I also think the new partition explainer could be useful for time series since it would account for the time ordering well while improving performance, but that would require building a masker that provides not just a masking call but also a `.clustering` to partition the inputs into a binary tree of related features (like recursively splitting by time).

**Reply by thedatadecoder** [reply]
Hi @slundberg , the link you shared seems to be expired. Can you please share a new link for example of custom masking function?

**Reply by suzuki-shm** [reply]
@thedatadecoder 
Hi, it would be https://github.com/slundberg/shap/blob/master/notebooks/api_examples/maskers/custom.ipynb .

**Reply by thedatadecoder** [reply]
Thanks a lot!!

**Reply by ricardobarroslourenco** [reply]
Hi @slundberg ! Is there any update on the masking for TS models?

> The short answer is that I think you should explain the impact of past time points, since that is often the biggest driving factor of the prediction. The tricky part about time series data is the strong time dependence makes realistically masking the features tricky. While it would be great to have, there are no time-series maskers built into SHAP right now. But you could use a custom masker and just impute forward or something. See here for an example of using a custom masking function in the new API: https://github.com/slundberg/shap/blob/master/notebooks/maskers/custom.ipynb
> 
> You could even go more all out and try to build a masker that is based on higher level features of the time series, like a DCT, or FWT transform. I also think the new partition explainer could be useful for time series since it would account for the time ordering well while improving performance, but that would require building a masker that provides not just a masking call but also a `.clustering` to partition the inputs into a binary tree of related features (like recursively splitting by time).

**Comment by luiztauffer** [comment]
following

**Comment by gideon-ai** [comment]
This is an interesting application of SHAP. Would love to see progress in this area.

**Reply by thedatadecoder** [reply]
You can use this to explain generalized additive regression models (Prophet is also a type of generalized additive model): https://shap.readthedocs.io/en/latest/example_notebooks/overviews/An%20introduction%20to%20explainable%20AI%20with%20Shapley%20values.html#Explaining-an-additive-regression-model
